src/main/java/org/kairosdb/core/aggregator/FilterAggregator.java

```diff
@@ -109,14 +109,25 @@ public FilterDataPointAggregator(DataPointGroup innerDataPointGroup) {
             }
         }
 
+        private boolean areTypesComparable(DataPoint dp, Class thresholdClass) {
+            return (String.class.isAssignableFrom(thresholdClass) && StringDataPoint.API_TYPE.equals(dp.getApiDataType())) ||
+                    (Number.class.isAssignableFrom(thresholdClass) && !StringDataPoint.API_TYPE.equals(dp.getApiDataType()));
+        }
+
         public boolean hasNext() {
             boolean foundValidDp = false;
             while (!foundValidDp && currentDataPoint != null) {
                 boolean filterOutValue = false;
-                if (Double.class.isAssignableFrom(m_threshold.getClass())) {
-                    filterOutValue = filterOut(m_filterop, currentDataPoint.getDoubleValue(), (Double) m_threshold);
-                } else if (StringDataPoint.API_TYPE.equals(currentDataPoint.getApiDataType())) {
-                    filterOutValue = filterOut(m_filterop, ((StringDataPoint) currentDataPoint).getValue(), m_threshold.toString());
+                if (areTypesComparable(currentDataPoint, m_threshold.getClass())) {
+                    if (Number.class.isAssignableFrom(m_threshold.getClass())) {
+                        filterOutValue = filterOut(m_filterop, currentDataPoint.getDoubleValue(), ((Number) m_threshold).doubleValue());
+                    } else if (StringDataPoint.API_TYPE.equals(currentDataPoint.getApiDataType())) {
+                        filterOutValue = filterOut(m_filterop, ((StringDataPoint) currentDataPoint).getValue(), m_threshold.toString());
+                    }
+                } else {
+                    // when types are different, we filter consider the datapoint to be non-equal to the threshold, so we just check if the operation is non-equal
+                    // for all other operations the datapoint will NOT be filtered out
+                    filterOutValue = FilterOperation.NE.equals(m_filterop);
```

Comment:
It's probably not such an important edge case, but I think that we might want to invert this, i.e. make it
```
filterOutValue = !FilterOperation.EQUAL.equals(m_filterop);
```

As it currently stands, doing a `GT` filter on numerical data points with a string threshold (or the same with string data points and numerical threshold) will allow all records through. I think that the general intention of e.g. a `GT` filter is "give me all values `x` where `(x > threshold) == true`". If x is a string and threshold is a double, then `x > threshold` can never true, so `x` should be dropped, but right now that `x` will be included in the returned results.

This also translates to inverting this filter downstream, so if we say that we want all data points where the `x > 100`, then we'll translate that into a filter of `x <= 100`, which would currently give back all numerical values that are less than equal to 100, as well as giving back all string values.

Reply:
Changed the code: if dataPoint type and threshold type are different, the datapoint *WILL* be filtered out